### PR TITLE
Timeout jlm server connection using elapsed time and nbr attempts

### PR DESCRIPTION
**Pull request by lumpfish**
attempts

Signed-off-by: Simon Rushton <rushton@uk.ibm.com>

### CODE REVIEW

**Comment by lumpfish** [comment]
Fixes https://github.com/AdoptOpenJDK/openjdk-systemtest/issues/370

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java [hunk 1]
```diff
@@ -171,32 +175,40 @@ private void getSecureServerConnection(String user, String pw) {
 		String[] credentials = new String[] { user, pw };
 		props.put("jmx.remote.credentials", credentials);             
 
-		// If we didn't connect keep try - the jvm being monitored might not be 
-		// ready to accept connections yet 
 		boolean connected = false;
 		int attempts = 0;
-		while (connected == false && attempts < 30) {        
+		long connectTimeout = 300L * 1000L * 1000L * 1000L;  // 300 secs in nanoseconds
```
Comment (karianna) [inline openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:180]: Duplicate code block?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java [hunk 2]
```diff
@@ -171,32 +175,40 @@ private void getSecureServerConnection(String user, String pw) {
...
+				connectElapsed = connectEnd - connectStart;
 
 				attempts++;           	
 				connected = false;
 
-				if (attempts == 30) {
-					Message.logOut("Failed to connect to Monitored VM after 30 attempts - giving up.  Connection Exception received is below:");	 
+				if (connectElapsed > connectTimeout || attempts == 30 ) {
+					Message.logOut("Failed to connect to Monitored VM after " + attempts + " attempts in " + (connectElapsed / 1000000000) + " seconds - giving up.  Connection Exception received is below:");	 
 					ie.printStackTrace();
```
Comment (karianna) [inline openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:200]: avoid printStackTrace?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java [hunk 3]
```diff
@@ -171,32 +175,40 @@ private void getSecureServerConnection(String user, String pw) {
...
+				connectEnd = System.nanoTime();
+				connectElapsed = connectEnd - connectStart;
 
 				attempts++;           	
 				connected = false;
 
-				if (attempts == 30) {
-					Message.logOut("Failed to connect to Monitored VM after 30 attempts - giving up.  Connection Exception received is below:");	 
+				if (connectElapsed > connectTimeout || attempts == 30 ) {
+					Message.logOut("Failed to connect to Monitored VM after " + attempts + " attempts in " + (connectElapsed / 1000000000) + " seconds - giving up.  Connection Exception received is below:");
```
Comment (karianna) [inline openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:199]: magic number could be extracted

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 1 above)
Comment (lumpfish) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:180]: Please can you be more specific about which "duplicate code block" you mean?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 2 above)
Comment (lumpfish) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:200]: Are you suggesting that the test should not execute `printStackTrace`?  Ensuring that it is executed was the purpose of this PR as described in https://github.com/AdoptOpenJDK/openjdk-systemtest/issues/370.

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 3 above)
Comment (lumpfish) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:199]: Which 'magic number' are you referring to, please?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 1 above)
Comment (karianna) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:180]:
```
long connectTimeout = 300L * 1000L * 1000L * 1000L;  // 300 secs in nanoseconds
			try {			long connectStart = System.nanoTime();
				Message.logOut("Trying to connect using JMXConnectorFactory");			long connectEnd = 0L;
		long connectElapsed = 0L;
		while (connected == false && attempts < 30 && (connectElapsed < connectTimeout) ) {
			Message.logOut("Attempting to connect");
			try {    
				connector = JMXConnectorFactory.connect(address);					connector = JMXConnectorFactory.connect(address);
				connected = true;					connected = true;
			} catch (IOException ie) {				} catch (IOException ie) {
				// If we didn't connect, keep trying - the JVM being monitored might not be 					// If we didn't connect keep trying - the jvm being monitored might not be 
				// ready to accept connections yet.					// ready to accept connections yet 
				try {           	  					connectEnd = System.nanoTime();
					Thread.sleep(10000);					connectElapsed = connectEnd - connectStart;
```

In ServerConnector.java line 112 and now here again?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 3 above)
Comment (karianna) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:199]: The 30 and the 1000000000 could both be extracted?

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 1 above)
Comment (lumpfish) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:180]: Thanks for the clarification.
The class has two similar methods - `getServerConnection()` and `getSecureServerConnection()` with much common code between them.  They could refactored to share much of that code, but that wasn't the focus of this PR.  They both attempt to establish a JMX connection with a max attempts (and now an elapsed timeout) retry capability.  With the code as it is, each method has its own value for attempts and timeout.  They could be the same (as they are now), but different values could in theory be set for the secure vs. non-secure connections.
(There are other ways the JLM tests might benefit from some refactoring too - there are other open issues which advocate that.)

File: openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java (hunk 3 above)
Comment (lumpfish) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:199]: Thanks for clarifying.  By 'extracted' you mean something like
```
maxAttempts = 30;
nanosecondsPerSecond = 1000L * 1000L * 1000L;
```
?

Comment (karianna) [reply openjdk.test.jlm/src/test.jlm/net/adoptopenjdk/test/jlm/remote/ServerConnector.java:199]: Something like that yes, it's probably just personal programming style preference, but makes it easier to parameterize/override later.

**Comment by lumpfish** [comment]
@karianna - I've refactored the code to get rid of the duplicate code between `getServerConnection()` and `getSecureServerConnection()`. Please take another look.

New test run is here: https://ci.adoptopenjdk.net/view/Test_grinder/job/Grinder/4267.

When we're ready to merge, please can we "Squash and merge".  No need to include my debugging and typo commits in the repository history.

**Review by karianna — APPROVED** [review]
LGTM

**Comment by karianna** [comment]
@lumpfish - I'm happy with it (thank you!) - save to merge during release day?

**Comment by lumpfish** [comment]
Merging now the Oct 2020 releases are finished.

**Pull request merged**